File: Documents/GitHub/Web-Projects/project5/parser.js
import { sine, cosine, tangent } from './calc_trig.js';
import { absolute, factorial, percentage, log, squareRoot, cubeRoot, exponent, is_prime }  from './calc_math.js';
import { dec_to_binary, dec_to_octal, dec_to_hex, F_to_C } from './calc_conversions.js';
import { mean, median, mode, maximum, find_smallest} from './calc_statistics.js';

// parses through the tokenized tokens
class Parser {

    advance(tokens) {
        const current_token = tokens.shift();
        return current_token;
    }
    
    parse_expr(tokens) {
        let result = this.parse_term(tokens);

        while (tokens[0] === '+' || tokens[0] === '-') {
            if (tokens[0] === '+') {
                this.advance(tokens);
                result += this.parse_term(tokens);
            } else if (tokens[0] === '-') {
                this.advance(tokens);
                result -= this.parse_term(tokens);
            }
        }

        return result;
    }
    
    parse_term(tokens) {
        let result = this.parse_factor(tokens);

        while (tokens[0] === '*' || tokens[0] === '/' || tokens[0] === '%') {
            if (tokens[0] === '*') {
                this.advance(tokens);
                result *= this.parse_factor(tokens);
            } else if (tokens[0] === '/') {
                this.advance(tokens);
                let divisor = this.parse_factor(tokens);

                if (divisor != 0) {
                    result = result * 1.0 / divisor;
                } else {
                    console.log("Cannot divide by 0");
                    return NaN;
                }

            } else if (tokens[0] === '%') {
                this.advance(tokens);
                let divisor = this.parse_factor(tokens);

                if (divisor != 0) {
                    result = result % divisor;
                } else {
                    console.log("Cannot divide by 0");
                }
            
            }
        }

        return result;
    }
    
    parse_factor(tokens) {
        if (tokens[0] === '(') {
            this.advance(tokens);
            let result = this.parse_expr(tokens);
            this.advance(tokens);
            return result;
    } else if (tokens[0] === '√' || tokens[0] === '∛' || tokens[0] === 'sin' || tokens[0] === 'cos' || tokens[0] === 'tan') {
        return this.parse_function(tokens);
    } else if (tokens[0] === 'log' || tokens[0] === 'exponent' || tokens[0] === 'percentage' || tokens[0] === 'temp' || tokens[0] === 'factorial' || tokens[0] === 'is_prime') {
        return this.parse_function(tokens);
    } else if (tokens[0] === 'odds' || tokens[0] === 'evens' || tokens[0] === 'squares' || tokens[0] === 'primes') {
        return this.parse_function(tokens);
    } else if (tokens[0] === 'absolute' || tokens[0] === 'to_binary' || tokens[0] === 'to_octal' || tokens[0] === 'to_hex' || tokens[0] === 'is_prime') {
        return this.parse_function(tokens);
    } else if (tokens[0] === 'π') {
        this.advance(tokens);
        return Math.PI;
    } else if (tokens[0] === 'mean' || tokens[0] === 'median' || tokens[0] === 'mode' || tokens[0] === 'min' || tokens[0] === 'max') {
        return this.parse_function(tokens);
    } else {
        return this.parse_digit_seq(tokens);
    }
}
    
    parse_function(tokens) {
        let func = tokens[0];

        this.advance(tokens);

        let argument = null;

        if (tokens[0] === '(') {
            this.advance(tokens);

            let arr = [];
            
            let base = this.parse_expr(tokens);
            arr.push(base);
            console.log("Parsing function:", func, "with base:", base);

            while (tokens[0] != ')') {
                if (tokens[0] === ',') {
                    if (func === 'mean' || func === 'median' || func === 'mode' || func === 'max' || func === 'min') {
                        this.advance(tokens);
                        let next_val = this.parse_expr(tokens);
                        arr.push(next_val);
                    } else {
                        this.advance(tokens);
                        argument = this.parse_expr(tokens);
                        console.log("Parsing function:", func, "with base:", base, "and argument:", argument);
                    }
                }
            }

            this.advance(tokens); // drop ')'

            // if (tokens[0] === ',') {
            //     this.advance(tokens);
            //     argument = this.parse_expr(tokens);
            //     console.log("Parsing function:", func, "with base:", base, "and argument:", argument);
            // }

            // if (tokens[0] === ')') {
            //     this.advance(tokens);
            // } 

            switch(func) {
                case 'sin':
                    return sine(base);
                case 'cos':
                    return cosine(base);
                case 'tan':
                    return tangent(base);
                case '√':
                    return squareRoot(base);
                case '∛':
                    return cubeRoot(base);
                case 'temp':
                    return F_to_C(base);
                case 'absolute':
                    return absolute(base);
                //case 'primes':
                    //return generate_primes(base);
                case 'factorial':
                    return factorial(base);
                case 'to_binary':
                    return dec_to_binary(base);
                case 'to_octal':
                    return dec_to_octal(base);
                case 'is_prime':
                    return is_prime(base);
                case 'mean':
                    return mean(arr);
                case 'median':
                    return median(arr);
                case 'max':
                    return maximum(arr);
                case 'min':
                    return find_smallest(arr);
                case 'mode':
                    return mode(arr);
                case 'to_hex':
                    return dec_to_hex(base);
                case 'log': 
                    // Ensure base and argument are valid for log calculation
                    if (base <= 0 || base == 1 || argument <= 0) {
                        console.log("Base and argument must be positive, and base cannot be 1 for log function");
                        break;
                    } else {
                        return log(base, argument); // Use custom log function
                    }
                case 'exponent':
                    return exponent(base, argument); // Use custom exponent function
                //case 'is_prime':
                    //return is_prime(base);
                case 'percentage':
                    return percentage(base, argument); // Use custom percentage function
                //case 'odds':
                    //return generate_odd_numbers(base, argument);
                //case 'evens':
                    //return generate_even_numbers(base, argument);
                //case'squares':
                    //return generate_squares(base, argument);
                //case 'isPrime':
                    //return is_prime(base);
                default: 
                    console.log('unknown function')
                    break;
            }
        }
    }

    parse_digit_seq(tokens) {
        let negative = false;
        if (tokens[0] === 'neg') {
            negative = true;
            this.advance(tokens);
        }

        let result = 0;

        if (tokens[0].includes('.')) {
            result = parseFloat(tokens[0]);
        } else {
            result = parseInt(tokens[0]);
        }

        this.advance(tokens);

        if (negative) {
            result = result * -1;
        }

        return result;
    }

}

export default Parser;

File: Documents/GitHub/Web-Projects/project2/parser.rb
require_relative 'methods'
# A class that is in charge of parsing tokens and evaluating expressions
class Parser
  include CalculatorMethods
  
  # Captures and sets up current token
  # @param tokens [Array<String>] The tokens representing a mathematical expression
  # @return [Parser] A new instance of Parser
  def initialize(tokens)
    @tokens = tokens
    @current_token = @tokens.shift
  end

  # Advances to the next token, like dequeue
  # This method updates the current token to the next token in the sequence
  # @return [void]
  def advance
    @current_token = @tokens.shift
  end

  # Parsing an entire expression (expr → term { + term | - term })
  # @return [Float, Integer] The result of the evaluated expression
  def parse_expr
    result = parse_term
    
    while @current_token == '+' || @current_token == '-'
      if @current_token == '+'
        advance
        result += parse_term
      elsif @current_token == '-'
        advance
        result -= parse_term
      end
    end
    result
  end

  # Parsing a term (term → factor { * factor | / factor | % factor})
  # @return [Float, Integer] The result of the term evaluation
  def parse_term
    result = parse_factor
    
    while @current_token == '*' || @current_token == '/' || @current_token == '%'

      if @current_token == '*'
        #if multiplier
        advance #eat *
        result *= parse_factor
      elsif @current_token == '/'
        #if divisor
        advance #eat 
        divisor = parse_factor
        #check for divide by 0
        if divisor!=0
          result=result * 1.0 / divisor
        else
          #returns error
          return Float::INFINITY
        end

      elsif @current_token == '%'
        advance # eat '%'
        divisor = parse_factor
        #check for mod by 0
        if divisor!=0
          result=result % divisor
        else
          #returns error
          result = Float::INFINITY
        end

      end
    end
    result
  end

  # Parsing a factor (factor → ( expr ) | digit-seq)
  # @return [Float, Integer] The result of the factor evaluation
  def parse_factor
    if @current_token == '('
      advance # consume '('
      result = parse_expr
      advance # consume ')'
      result
    elsif @current_token == '√' || @current_token == '∛' || @current_token == 'sin' || @current_token == 'cos' || @current_token == 'tan' 
      parse_function
    elsif @current_token == 'log' || @current_token == 'exponent' || @current_token == 'percentage' || @current_token == 'temp' || @current_token == 'factorial'  
      parse_function
    elsif @current_token == 'odds' || @current_token == 'evens' || @current_token == 'squares' || @current_token == 'primes' || @current_token == 'isPrime'
      parse_function
    elsif @current_token == 'absolute' || @current_token == 'to_binary' || @current_token == 'to_octal' || @current_token == 'to_hex' || @current_token == 'is_prime'
      parse_function
    elsif @current_token == 'π'
      advance # consume 'π'
      Math::PI
    else
      parse_digit_seq
    end
  end

  # Parsing a function call (function → function-name ( expr { , expr }* ))
  # @return [Float, Integer] The result of the function evaluation
  def parse_function
    function = @current_token
    advance # consume the function name
  
    if @current_token == '('
      advance # consume '('
      base = parse_expr # First argument (base)
      
      if @current_token == ',' # Check if there is a second argument
        advance # consume ','
        argument = parse_expr # Second argument
      else
        argument = nil # No second argument
      end
  
      if @current_token == ')' # Check for closing parenthesis
        advance # consume ')'
      else
        return Float::INFINITY
      end
  
      case function
      when 'sin'
        sine(base) # Takes in radians
      when 'cos'
        cosine(base) # Takes in radians
      when 'tan'
        tangent(base) # Takes in radians
      when '√'
        square_root(base)
      when '∛'
        cube_root(base)
      when 'temp'
        fToC(base)
      when 'absolute'
        absolute(base)
      when 'primes'
        generate_primes(base)
      when 'factorial'
        factorial(base)
      when 'to_binary'
        decimal_to_binary(base)
      when 'to_octal'
        decimal_to_octal(base)
      when 'to_hex'
        decimal_to_hex(base)
      when 'log'
        # Ensure base and argument are valid for log calculation
        if base <= 0 || base == 1 || argument <= 0
          raise "Base and argument must be positive, and base cannot be 1 for log function"
        end
        log(base, argument) # Use custom log function
      when 'exponent'
        exponent(base, argument) # Use custom exponent function
      when 'is_prime'
        is_prime(base)
      when 'percentage'
        percentage(base, argument) # Use custom percentage function
      when 'odds'
        generate_odd_numbers(base, argument)
      when 'evens'
        generate_even_numbers(base, argument)
      when 'squares'
        generate_squares(base, argument)
      when 'isPrime'
        is_prime(base)
      else
        raise "Unknown function: #{function}"
      end
    else
      raise "Expected '(' after function name"
    end
  end
  

  # Parsing a sequence of digits (digit-seq → digit { digit })
  # @return [Float, Integer] The numerical value of the digit sequence
  def parse_digit_seq 

    negative= false
    if @current_token == 'q'
      #if the number should be negative, indicated by 'q', negative=true and eat the q token
      negative = true
      advance
    end

    if @current_token && @current_token.include?('.') 
      #if its a decimal
      result = @current_token.to_f
    else 
      result= @current_token.to_i
    end
    
    advance

    if negative
      result *= -1
    end

    result
  end
end